### PR TITLE
refactor(Ready): add pub getters and make fields private

### LINKED ISSUE #90: Ready is Ready-Only, should only have getters.
The `Ready` struct should not allow values to be mutated, it should only provide immutable getters.

https://github.com/pingcap/raft-rs/blob/1e0741a9aea3b242d33dadc46245ad611e7e7aa4/src/raw_node.rs#L84-L121

It's possible to do this by writing functions that return immutable borrows of the contained values, or using a crate like https://crates.io/crates/getset which will write them for you.

### PR DESCRIPTION
**Pull request by Ryan-Git**
Fixes #90

### CODE REVIEW

File: src/raw_node.rs [hunk 1]
```diff
@@ -174,6 +152,68 @@ impl Ready {
...
+    #[inline]
+    pub fn snapshot(&self) -> &Snapshot {
+        &self.snapshot
+    }
+
+    /// CommittedEntries specifies entries to be committed to a
+    /// store/state-machine. These have previously been committed to stable
+    /// store.
+    #[inline]
+    pub fn committed_entries(&self) -> Option<&[Entry]> {
```
Comment (BusyJay) [inline src/raw_node.rs]: So users have to clone the memory to handle committed entries in other thread now?

File: src/raw_node.rs [hunk 2]
```diff
@@ -174,6 +152,68 @@ impl Ready {
...
+        self.committed_entries.as_ref().map(|v| v.as_slice())
+    }
+
+    /// Messages specifies outbound messages to be sent AFTER Entries are
+    /// committed to stable storage.
+    /// If it contains a MsgSnap message, the application MUST report back to raft
+    /// when the snapshot has been received or has failed by calling ReportSnapshot.
+    #[inline]
+    pub fn messages(&self) -> &[Message] {
```
Comment (BusyJay) [inline src/raw_node.rs]: Channel usually move the message. So a clone seems unavoidable.

**Comment by Hoverbear** [comment]
Thanks for taking up the flag on this @Ryan-Git. :)

We will definitely need to include this in a non-patch version since it's an API break.

I think @BusyJay has a very interesting point:

> And what if users want to advance ready partially?

@BusyJay, could you expand more on this? I was unaware of this use.

**Comment by BusyJay** [comment]
For example, we may handle committed entries and entries differently so may not advance them all together. Maybe we need a more clear `Ready` mechanism for this. For example, Use `Ready` to represent readiness, and use `HandleProgress`, which is just an example name, to represent the handle progress. So `Ready` is free to be moved and modify, raft is only required to check the progress.

**Comment by Hoverbear** [comment]
@BusyJay Maybe we should revert the change on these particular fields, and merge this PR, then investigate how to move forward with this solution in another issue?

**Comment by Ryan-Git** [comment]
I could revert changes on messages and committed_entries mentioned above.

**Comment by Hoverbear** [comment]
@Ryan-Git Yeah, let's revert them for now. Thanks so much!

**Comment by Ryan-Git** [comment]
updated

**Comment by siddontang** [comment]
PTAL @BusyJay

File: tests/cases/test_raw_node.rs [hunk 3]
```diff
@@ -59,21 +59,22 @@ fn conf_change(t: ConfChangeType, node_id: u64) -> ConfChange {
     cc
 }
 
-fn new_ready(
```
Comment (Hoverbear) [inline tests/cases/test_raw_node.rs]: Maybe we can `#[derive(PartialEq)]` instead?

Comment (Hoverbear) [reply tests/cases/test_raw_node.rs]: Ah, nevermind. :)

Comment (Ryan-Git) [reply tests/cases/test_raw_node.rs]: emm.. do you mean to add the constructor or just keep it as it is?

**Comment by Ryan-Git** [comment]
> LGTM. Please resolves conflicts.
resolved

**Comment by Ryan-Git** [comment]
> LGTM. Please resolves conflicts.

resolved

**Comment by Hoverbear** [comment]
@Ryan-Git Thanks, I'm going to drive this to merge. :) I decided to use your PR to test our bors and I have a small hiccup. If you get some notifications you can just ignore them. Sorry about that!

**Comment by Hoverbear** [comment]
bors retry

**Comment by Hoverbear** [comment]
:tada: Thank you @Ryan-Git for your patience and hard work, and @bors[bot] for the merge. :wink:

**Comment by Ryan-Git** [comment]
@Hoverbear My pleasure :-)

**Pull request merged**